ext/oboe_metal/src/profiling.cc

```diff
@@ -0,0 +1,405 @@
+        return Qfalse;
+    } catch (...) {
+        string msg = "Exception in Profiling::profiling_run, can't recover, profiling shutting down";
+        OBOE_DEBUG_LOG_ERROR(OBOE_MODULE_RUBY, msg.c_str());
+        Profiling::shut_down();
+        return Qfalse;
+    }
+
+    return Qfalse;
+}
+
+// in case C++ misbehaves we will stop profiling
+// to be used when catching exceptions
+void Profiling::shut_down() {
+    static bool started = false;
+    // stop all profiling, the last one also stops the timer/signals
+    if (!started) {
+        for (pair<const pid_t, prof_data_t> &ele : prof_data_map) {
+            started = true;
+            profiling_stop(ele.first);
+        }
+    }
+    // avoid running any more profiling
+    profiling_shutdown = true;
+}
+
+VALUE Profiling::getTid() {
+    pid_t tid = AO_GETTID;
+
+    return INT2NUM(tid);
+}
+
+static void
+prof_atfork_prepare(void) {
+    // cout << "Parent getting ready" << endl;
+}
+
+static void
+prof_atfork_parent(void) {
+    // cout << "Parent let child loose" << endl;
+}
+
+// make sure new processes have a clean slate for profiling
+static void
+prof_atfork_child(void) {
+    // cout << "A child is born" << endl;
+    Frames::clear_cached_frames();
+    prof_data_map.clear();
+    running = false;
+
+    // make sure it has a timer ready, it is a per-process-timer
+    Profiling::create_timer();
+}
+
+void Profiling::create_timer() {
+    struct sigevent sev;
+
+    sev.sigev_value.sival_ptr = &timerid;
+    sev.sigev_notify = SIGEV_SIGNAL; /* Notify via signal */
+    sev.sigev_signo = SIGRTMAX;      /* Notify using this signal */
+
+    if (timer_create(CLOCK_REALTIME, &sev, &timerid) == -1) {
+        OBOE_DEBUG_LOG_ERROR(OBOE_MODULE_RUBY, "timer_create() failed");
+        profiling_shutdown = true;  // no profiling without clock
+    }
+}
+
+extern "C" void Init_profiling(void) {
+    Profiling::create_timer();
+    Frames::reserve_cached_frames();
+
+    // creates Ruby Module: AppOpticsAPM::CProfiler
+    static VALUE rb_mAppOpticsAPM = rb_define_module("AppOpticsAPM");
+    static VALUE rb_mCProfiler = rb_define_module_under(rb_mAppOpticsAPM, "CProfiler");
+
+    rb_define_singleton_method(rb_mCProfiler, "get_interval", reinterpret_cast<VALUE (*)(...)>(Profiling::get_interval), 0);
+    rb_define_singleton_method(rb_mCProfiler, "set_interval", reinterpret_cast<VALUE (*)(...)>(Profiling::set_interval), 1);
+    rb_define_singleton_method(rb_mCProfiler, "run", reinterpret_cast<VALUE (*)(...)>(Profiling::profiling_run), 1);
+    rb_define_singleton_method(rb_mCProfiler, "get_tid", reinterpret_cast<VALUE (*)(...)>(Profiling::getTid), 0);
+
+    pthread_atfork(prof_atfork_prepare,
+                   prof_atfork_parent,
+                   prof_atfork_child);
+
+    // TODO better understand the gc marking
+    // ____ needed???
+    // ____ does it last forever or is it reset after a gc?
+    for (int i = 0; i < BUF_SIZE; i++) rb_gc_mark(frames_buffer[i]);
```

Comment:
Does this need more attention based on the comment or is it low risk?

Reply:
I have 3 things I need to better understand and this is one of them ... WIP ...

Reply:
👍🏼